### PR TITLE
test: TableCollection.tsx (Frontend) - Dividing into smaller components and creating tests

### PR DESCRIPTION
**Pull request by yardz**
### SUMMARY
The purpose of this PR is to improve the readability and maintainability of component `src/components/dataViewCommon/TableCollection.tsx`.

#### Observed issues:
High complexity, makes the code more difficult to understand and to test.
**Proposed solution:** Divide the component into smaller, more specialized components.

#### Solution description
The component was initially divided into 2:  `TableCollectionHead` and `TableCollectionBody`
As the table's data model is complex, the tests were intended to just ensure that the pros were being send correctly to these components.

The `TableCollectionHead` is intended to render the table's head (`<thead />`). For this, it was divided into 2. The `TableCollectionHead` itself that takes care of the re-rendering of the lines (`<tr>`) and the `TableCollectionHeadCol` which takes care of the rendering of the cells (`<td>`).
The testing strategy of TableCollectionHead is also just to ensure that an object that minimally respects the interface of the `react-table` and sends the correct props to `TableCollectionHeadCol`.
The`TableCollectionHeadCol` test targets all states/behaviors variations supported by the component.

The component `TableCollectionBody` was also initially divided into 2 components `TableCollectionLoading` and `TableCollectionBodyContent`.
`TableCollectionLoading` is responsible for rendering when the table is in the loading state and has no information. The testing strategy is to cover the behaviors and ensure that the props are sent correctly to the component `TableCollectionCell`
The component `TableCollectionBodyContent` is responsible for rendering the data (the rows / `<tr>`) of the table. The testing strategy also consists of covering possible behaviors and sending the correct props to `TableCollectionCell`
`TableCollectionCell` is responsible for rendering a table cell (`<td>`), either containing data or in a loading state. The testing strategy consists of verifying that when receiving the cell data (props), it will behave properly.

In summary, `TableCollection`, `TableCollectionHead`, `TableCollectionBody`, `TableCollectionLoading` and `TableCollectionBodyContent` are responsible for sending the pops to a component below in your tree. Their tests are focused on ensuring that the props are being sent correctly.
`TableCollectionHeadCol` and `TableCollectionHeadCol` are responsible for render the table cells (`<td>/<th>`). The testing of these components is focused on behavior and correct rendering.

### TEST PLAN
No changes in behavior should occur and all tests must pass.

### CODE REVIEW

**Comment by yardz** [comment]
@ktmud, @nytai, @michael-s-molina @geido  If you can look, I appreciate it. I am still finding the tests very weak, if you have any suggestions on how to make them more robust please tell me.

**Comment by michael-s-molina** [comment]
Hey @yardz! Nice refactoring!

About the tests:

Try to access the elements by `role` instead of `test-id` which is not accessible to the users. You can use one of the [predefined roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Table_Role) `rowgroup`, `row`, `columnheader`, and `cell`.

With testing library, you're encouraged to test the user's interaction with the screen. Relying on the React component data like state, props or attributes leads to brittle tests that depend on implementation detail and don't provide as much value to users. You can review some of your expects of type `toHaveAttribute`. Think about the scenario where we change the implementation of the component (remove attributes or classes) but keep the same functionality to the user. Your tests should not fail after the change.

Instead of test for attributes:

```
render(<TableCollection {...} />);
expect(screen.getByTestId('loading')).toHaveAttribute(
    'data-loading',
    'loading',
);
```

Try to change the focus more to how the **user** will perceive the different component states.

```
render(<TableCollection loading {...} />);
const status = screen.getByRole('status');
expect(status).toBeInTheDocument();
expect(status).toHaveTextContent('Loading');
```

You're are finding the tests very weak because this component is a simple layer on top of `react-table` without too many additional behaviors. The majority of tests have already been made by the `react-table` team. This is common in this kind of component. You will find more complex test scenarios in components that have many custom behaviors or components created from scratch.

**Comment by yardz** [comment]
> Try to access the elements by role instead of test-id which is not accessible to the users. You can use one of the predefined roles rowgroup, row, columnheader, and cell.

These `role` are when you simulate a table with other elements (for example, with `<div>`). For tables, you should not place a same `<table role="table">`.
The link you sent says this: "The table value of the ARIA role attribute identifies the element containing the role as having a non-interactive table structure containing data arranged in rows and columns, similar to the native `<table>` HTML element." and this "Realize, it is much simpler to use the `<table>` element, along with all the related semantic elements and attributes that are all supported by all assistive technologies."

Specifically in this case, adding the `role` does not seem like a good option/best practices.


> Try to change the focus more to how the **user** will perceive the different component states.

The ideal was in fact to do this, and to proceed as an "integration test". However, `react-tables` are very complex, so creating the data for the test is extremely complex (so it is very easy to get it wrong).

As this component only calls two other components, I followed the unit testing strategy, where I only check the props. And emphasizing this component has no "interaction with the user" only display.

**Comment by michael-s-molina** [comment]
> Specifically in this case, adding the `role` does not seem like a good option/best practices.

@yardz I think you misunderstood what I said. The roles are already available to you because you are using a native HTML table. It's similar to `<button>`, you don't need to explicitly add a role. 

Anytime you want to know which roles are available you can always use [logRoles](https://testing-library.com/docs/dom-testing-library/api-accessibility/#logroles) helper function. 

If you execute this in `TableCollection.test.tsx` you will see that the roles are already there.
```
import { render, screen, logRoles } from 'spec/helpers/testing-library';
...
const { container } = render(<TableCollection {...(props as any)} />);
logRoles(container);
...
```

**Comment by yardz** [comment]
> If you execute this in `TableCollection.test.tsx` you will see that the roles are already there.

Very good! I didn't know that, thanks for the tip!

**Comment by nytai** [comment]
Do people actually find this re-implementation more readable? Personally, when I think of a table I think of it in its entirety, rarely am I thinking about the header and body as 2 completely separate entities, and a header row separately from a header block. When making changes I like to be able to quickly spot the difference in treatment between a header row and a body row, as they are often related. Additionally, this component was mostly a presentational layer on top of `react-table` and there wasn't much logic beyond a mapping of the objects provided by `react-table` into html table markup. Looking at the updated code it looks like some sub components are held in react memo hooks which makes me think there is more complex logic involved in the subcomponents and updating is costly. While these changes likely made unit testing of the subcomponents easier, I can't say readability has improved or that complexity has reduced, from my perspective.

**Comment by yardz** [comment]
> Personally, when I think of a table I think of it in its entirety, rarely am I thinking about the header and body as 2 completely separate entities, and a header row separately from a header

I agree with that, when using a table we really think of all the set. When using a table component, it doesn't really make sense to call the table, header, body separately. We need to send the header and the data and the "magic" must happen.

But in this case it is the construction of the table and not the use of a table (other components will use that table). So, in my view, this division makes sense. It is also worth mentioning that this component has a behavior logic (like props loading) and some functions to add different props on html entities (like getTableProps and getTableBodyProps).

### Tests

The goal here was to create tests. The end result was 7 test files with an average of 100 test lines in each and not all behavior scenarios are covered (only the most important ones). If we put all the tests into a single component, we would have an archive of approximately 700 test lines, something that is not easy to maintain in long term.

In general, I create tests and components with TDD, so large and complex components need to be broken to be workable.

### Complexity

About complexity. The final components are less complex. When I refer to complexity I mean cyclomatic complexity.

Code with less [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) is easier to understand because it is simpler. A person who is unaware of react-tables (like me - I’ve never seen this before) and is not extremely familiar with the old component you will not be able to understand.
I spent a few hours myself to understand what was being done there... And I only started to move forward when I started to delete code to see what happened. It was not an easy task at all.

### Clean Code

Of course, you know. But [this link](https://gist.github.com/wojteklu/73c6914cc446146b8b533c0988cf8d29) has a short list of general topics that are an excellent guide for day to day. Basically one of the general rules is `Keep it simple stupid. Simpler is always better. Reduce complexity as much as possible.` (personally is my favorite).


### useMemo
I saw no need, but since the original component was memoized `export default React.memo` I preferred to keep something in the same way.
Anyway, I don't think that useMemo makes reading difficult to deserve attention.

@nytai If even with these arguments you are not convinced I can close the PR and return with the old code, and do the tests that I can.
